contracts/TuiChainLoanManagement.sol

```diff
@@ -0,0 +1,368 @@
+            _requestedValueDaiWei);
+
+        feeRecipient           = _feeRecipient;
+        loanRecipient          = _loanRecipient;
+        expirationTime         = block.timestamp.add(_secondsToExpiration);
+        fundingFeeDaiWeiPerDai = _fundingFeeDaiWeiPerDai;
+        paymentFeeDaiWeiPerDai = _paymentFeeDaiWeiPerDai;
+        requestedValueDai      = _requestedValueDai;
+
+        phase          = Phase.Funding;
+        loanToken      = TuiChainLoanToken(0);
+        fundedDai      = 0;
+        paidDai        = 0;
+        daiWeiPerToken = 0;
+    }
+
+    function _requirePositiveIntegerDai(
+        uint256 _daiWei) private pure returns (uint256 _dai)
+    {
+        require(
+            _daiWei > 0 && _daiWei.mod(1e18) == 0,
+            "Must be a positive, integer amount of Dai (multiple of 10^18)");
+
+        return _daiWei.div(1e18);
+    }
+
+    function _updatePhase(Phase _newPhase) private
+    {
+        assert(phase != _newPhase);
+
+        phase = _newPhase;
+        emit PhaseUpdate(_newPhase);
+    }
+
+    function _tryExpire() private
+    {
+        if (phase == Phase.Funding && block.timestamp >= expirationTime)
+            _updatePhase(Phase.Expired);
+    }
+
+    /**
+     * Expire the loan if the funding deadline has passed.
+     *
+     * Fails if current phase is not Funding or Expired.
+     *
+     * @return _expired true if the loan becomes or already was expired, false
+     *     otherwise
+     */
+    function checkExpiration() public returns (bool _expired)
+    {
+        _tryExpire();
+
+        require(phase == Phase.Funding || phase == Phase.Expired);
+
+        return phase == Phase.Expired;
+    }
+
+    /**
+     * Deposit some Dai to fund the loan.
+     *
+     * Fees are added to _valueDaiWei. The actual transferred value, in Dai Wei,
+     * is given by:
+     *
+     *     _valueDaiWei + fundingFeeDaiWeiPerDai * (_valueDaiWei / 1e18)
+     *
+     * @param _valueDaiWei Amount of funds to provide (before fees) in Dai Wei
+     */
+    function provideFunds(uint256 _valueDaiWei) external
+    {
+        _tryExpire();
+
+        // checks
+
+        require(phase == Phase.Funding);
+
+        uint256 valueDai = _requirePositiveIntegerDai(_valueDaiWei);
+        require(fundedDai.add(valueDai) <= requestedValueDai);
+
+        // effects
+
+        emit FundsUpdate(
+            fundedDai.mul(1e18), fundedDai.add(valueDai).mul(1e18));
+
+        fundedDai = fundedDai.add(valueDai);
+        unclaimedTokens[msg.sender] = unclaimedTokens[msg.sender].add(valueDai);
+
+        if (fundedDai == requestedValueDai)
+        {
+            _updatePhase(Phase.Active);
+
+            loanToken = new TuiChainLoanToken(requestedValueDai);
+        }
+
+        // interactions
+
+        uint256 feeDaiWei = fundingFeeDaiWeiPerDai.mul(valueDai);
+
+        dai.safeTransferFrom(
+            msg.sender, address(this), _valueDaiWei.add(feeDaiWei));
```

Comment:
This is even possible? This transfers more dai than the investor want.

Reply:
This behavior is noted in the function documentation, although it may be a bit counterintuitive.

The alternative would be for the caller to provide the value with fees included. However, we would then need to compute the fee by dividing the value by fundingFeeDaiWeiPerDai, which could easily not lead to an exact result.

This could be avoided by requiring the input value to be divisible by fundingFeeDaiWeiPerDai, although i'd argue that that behavior would be even more confusing than the current one.

Reply:
The thing is, how to transfer an amount of DAI (the fee) that may not exist? Is this the desired behavior?

Reply:
There are as many guarantees that the caller has enough balance to cover _valueDaiWei as there are that he has enough to cover _valueDaiWei + feeDaiWei. This is the desired behavior.